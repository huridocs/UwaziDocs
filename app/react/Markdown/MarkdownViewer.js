import PropTypes from 'prop-types';
import React, { Component } from 'react';
import markdownIt from 'markdown-it';
import rison from 'rison';
import CustomHookComponents from './CustomHooks';
import CustomComponents from './components';

const placeholder = '{-CUSTOMCOMPONENT-}';
const componentSubstitution = '{-spliter-}{-CUSTOMCOMPONENT-}{-spliter-}';
const spliter = '{-spliter-}';
const customComponentMatcher = /{\w+}\(.+\)\(.+\)|{\w+}\(.+\)/g;
const md = markdownIt();

export class MarkdownViewer extends Component {
  static errorHtml(index) {
    return (
      <p key={index} className="error">
        <br />
        <strong><i>Custom component markup error: unsuported values! Please check your configuration</i></strong>
        <br />
      </p>
    );
  }

  static customHook(config, index) {
    let output;
    try {
      const props = rison.decode(config);
      if (!CustomHookComponents[props.component]) {
        throw new Error('Invalid  component');
      }
      const Element = CustomHookComponents[props.component];
      output = <Element {...props} key={index} />;
    } catch (err) {
      output = MarkdownViewer.errorHtml(index);
    }
    return output;
  }

  list(config, index) {
    const listData = this.props.lists[this.renderedLists] || {};
    const output = <CustomComponents.ItemList key={index} link={`/library/${listData.params}`} items={listData.items} options={listData.options}/>;
    this.renderedLists += 1;
    return output;
  }

  customComponent(customComponentString, index) {
    const customComponentTypeMatcher = /{(.+)}\(/;
    const customComponentOptionsMatcher = /{\w+}(\(.+\)\(.+\))|{\w+}(\(.+\))/g;
    const type = customComponentString.match(customComponentTypeMatcher)[1];
    const configMatch = customComponentOptionsMatcher.exec(customComponentString);
    let config;
    if (configMatch) {
      config = configMatch[1] || configMatch[2];
    }

    if (type === 'customhook') {
      return MarkdownViewer.customHook(config, index);
    }

    if (['vimeo', 'youtube', 'media'].includes(type)) {
      return <CustomComponents.MarkdownMedia key={index} config={config} />;
    }

    if (type === 'list') {
      return this.list(config, index);
    }

    return <div key={index}>{customComponentString}</div>;
  }

  markdownToHtml(markdown) {
    const customComponents = markdown.match(customComponentMatcher);
    const htmlChunks = markdown.replace(customComponentMatcher, componentSubstitution).split(spliter);
    let customComponentsPrinted = 0;
    return htmlChunks.map((chunk, index) => {
      if (chunk === placeholder) {
        const component = this.customComponent(customComponents[customComponentsPrinted], index);
        customComponentsPrinted += 1;
        return component;
      }
      const __html = md.render(chunk);
      return <div key={index} dangerouslySetInnerHTML={{ __html }}/>; // eslint-disable-line
    });
  }

  render() {
    this.renderedLists = 0;
    return (
      <div className="markdown-viewer">
        {this.markdownToHtml(this.props.markdown).map(item => item)}
      </div>
    );
  }
}

MarkdownViewer.defaultProps = {
  lists: [],
  markdown: ''
};

MarkdownViewer.propTypes = {
  markdown: PropTypes.string,
  lists: PropTypes.arrayOf(PropTypes.object)
};

export default MarkdownViewer;
