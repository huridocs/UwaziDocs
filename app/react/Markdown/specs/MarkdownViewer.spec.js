import React, { Component } from 'react';

import { shallow } from 'enzyme';

import CustomHookComponents from '../CustomHooks';
import MarkdownViewer from '../MarkdownViewer';

describe('MarkdownViewer', () => {
  let component;
  let props;

  beforeEach(() => {
    props = {
      markdown: '## MarkdownContent'
    };
  });

  const render = () => {
    component = shallow(<MarkdownViewer {...props} />);
  };

  describe('render', () => {
    describe('when markdown is empty', () => {
      it('should not render anything', () => {
        props.markdown = '';

        render();
        expect(component).toMatchSnapshot();
      });
    });

    it('should render markdown', () => {
      props.markdown = `## title\nSome text with a [URL](http://google.com) inside.
        \n\nWhich should be in its own line, separated with TWO line breaks (to create a new <p> Element)
        \n\n\`\`\`javascript\nCode\n\`\`\`
      `;

      render();
      expect(component).toMatchSnapshot();
    });

    it('should support containers with custom classNames', () => {
      props.markdown = '::: test classname \ntext inside a div';

      render();
      expect(component).toMatchSnapshot();
    });

    it('should render media components', () => {
      props.markdown = `{vimeo}(1234, options)
      \n\n{youtube}(1234)
      \n\n{media}(config options)
      `;

      render();
      expect(component).toMatchSnapshot();
    });

    it('should render list components', () => {
      props.lists = [
        { params: 'param1', items: ['items1'], options: { option1: 'optionValue' } },
        { params: 'param2', items: ['items2'] },
        { params: 'param3', items: ['items3'] }
      ];

      props.markdown = `{list}(first list)
        \n\n{list}(http://secondList)(limit:3)
        \n\n{list}(asd)
      `;

      render();
      expect(component).toMatchSnapshot();
    });

    it('should render customHook components and show an error for invalid ones', () => {
      CustomHookComponents.validcomponent = class customComponent extends Component {
        render() {
          return <p>Custom</p>;
        }
      };

      props.markdown = `
        \n\n{customhook}(component:validcomponent,prop:'a prop',array:!((child:!(a),bool:!t),(child:!(b))))
        \n\nShould allow markdown between hooks
        \n\n{customhook}(component:validcomponent)
        \n\n{customhook}(component:invalidcomponent)
      `;

      render();
      expect(component).toMatchSnapshot();
    });

    it('should be able to render properly custom components not separated by \\n\\n', () => {
      props.markdown = '{youtube}(youtubeUrl)\n{media}(mediaUrl)';

      render();
      expect(component).toMatchSnapshot();
    });

    it('should not render html by default', () => {
      props.markdown = '<div><h1>should be all a escaped string</h1></div>';

      render();
      expect(component).toMatchSnapshot();
    });

    it('should render Link', () => {
      props.markdown = '{link}(url:this_is_the_url,label:\'this is a label\')'
        + '\n<MarkdownLink url="the_url">label</MarkDownLink>';
      props.html = true;

      render();
      expect(component).toMatchSnapshot();
    });

    describe('when passing html true prop', () => {
      it('should render html', () => {
        props.html = true;
        props.markdown = '<div><h1>test</h1></div>';

        render();
        expect(component).toMatchSnapshot();
      });

      it('should render customComponents nested inside html', () => {
        props.html = true;
        props.lists = [
          { params: 'param1', items: ['items1'], options: { option1: 'optionValue' } },
        ];
        props.markdown = '<div>{list}(url)(options) <div>{youtube}(url)</div></div>';

        render();
        expect(component).toMatchSnapshot();
      });
    });

    describe('when not valid html', () => {
      it('should not fail', () => {
        props.markdown = '<div><h</div>';
        props.html = true;

        render();
        expect(component).toMatchSnapshot();
      });
    });

    it('should render a searchbox', () => {
      props.markdown = '{searchbox}(placeholder:\'this is a placeholder\')';

      render();
      expect(component).toMatchSnapshot();
    });
  });
});
