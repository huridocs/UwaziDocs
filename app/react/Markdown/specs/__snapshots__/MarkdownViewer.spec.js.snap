// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkdownViewer render should be able to render properly custom components not separated by \\n\\n 1`] = `
<div
  className="markdown-viewer"
>
  <MarkdownMedia
    config="(youtubeUrl)"
    key="0"
  />
  

  <MarkdownMedia
    config="(mediaUrl)"
    key="2"
  />
  

</div>
`;

exports[`MarkdownViewer render should not render html by default 1`] = `
<div
  className="markdown-viewer"
>
  <p
    key="0"
  >
    &lt;div&gt;&lt;h1&gt;should be all a escaped string&lt;/h1&gt;&lt;/div&gt;
  </p>
  

</div>
`;

exports[`MarkdownViewer render should remove Dataset tags 1`] = `
<div
  className="markdown-viewer"
>
  <p
    key="0"
  >
    test
  </p>
  

  

  <SearchBox
    classname=""
    key="4"
    placeholder=""
  />
  

  

  <div
    key="8"
  >
    test
  </div>
</div>
`;

exports[`MarkdownViewer render should render Link 1`] = `
<div
  className="markdown-viewer"
>
  <MarkdownLink
    classname=""
    key="0"
    label="this is a label"
    url="this_is_the_url"
  />
  

  <p
    key="2"
  >
    <MarkdownLink
      classname=""
      key="0"
      url="the_url"
    >
      label
    </MarkdownLink>
  </p>
  

</div>
`;

exports[`MarkdownViewer render should render a searchbox 1`] = `
<div
  className="markdown-viewer"
>
  <SearchBox
    classname=""
    key="0"
    placeholder="this is a placeholder"
  />
  

</div>
`;

exports[`MarkdownViewer render should render customHook components and show an error for invalid ones 1`] = `
<div
  className="markdown-viewer"
>
  <customComponent
    array={
      Array [
        Object {
          "bool": true,
          "child": Array [
            "a",
          ],
        },
        Object {
          "child": Array [
            "b",
          ],
        },
      ]
    }
    component="validcomponent"
    key="0"
    prop="a prop"
  />
  

  <p
    key="2"
  >
    Should allow markdown between hooks
  </p>
  

  <customComponent
    component="validcomponent"
    key="4"
  />
  

  <p
    className="error"
    key="6"
  >
    <br />
    <strong>
      <i>
        Custom component markup error: unsuported values! Please check your configuration
      </i>
    </strong>
    <br />
    Invalid  component
    <br />
  </p>
  

</div>
`;

exports[`MarkdownViewer render should render list components 1`] = `
<div
  className="markdown-viewer"
>
  <ItemList
    items={
      Array [
        "items1",
      ]
    }
    key="0"
    link="/library/param1"
    options={
      Object {
        "option1": "optionValue",
      }
    }
  />
  

  <ItemList
    items={
      Array [
        "items2",
      ]
    }
    key="2"
    link="/library/param2"
    options={Object {}}
  />
  

  <ItemList
    items={
      Array [
        "items3",
      ]
    }
    key="4"
    link="/library/param3"
    options={Object {}}
  />
  

</div>
`;

exports[`MarkdownViewer render should render markdown 1`] = `
<div
  className="markdown-viewer"
>
  <h2
    key="0"
  >
    title
  </h2>
  

  <p
    key="2"
  >
    Some text with a 
    <a
      href="http://google.com"
      key="1"
    >
      URL
    </a>
     inside.
  </p>
  

  <p
    key="4"
  >
    Which should be in its own line, separated with TWO line breaks (to create a new &lt;p&gt; Element)
  </p>
  

  <pre
    key="6"
  >
    <code
      className="language-javascript"
      key="0"
    >
      Code

    </code>
  </pre>
  

</div>
`;

exports[`MarkdownViewer render should render media components 1`] = `
<div
  className="markdown-viewer"
>
  <MarkdownMedia
    config="(1234, options)"
    key="0"
  />
  

  <MarkdownMedia
    config="(1234)"
    key="2"
  />
  

  <MarkdownMedia
    config="(config options)"
    key="4"
  />
  

</div>
`;

exports[`MarkdownViewer render should render properly a selfclosing XML tags 1`] = `
<div
  className="markdown-viewer"
>
  <p
    key="0"
  >
    test
  </p>
  

  <SearchBox
    classname=""
    key="2"
    placeholder=""
  />
  

  <div
    key="4"
  >
    test
  </div>
</div>
`;

exports[`MarkdownViewer render should support containers with custom classNames 1`] = `
<div
  className="markdown-viewer"
>
  <div
    className="test classname"
    key="0"
  >
    <p
      key="0"
    >
      text inside a div
    </p>
    

  </div>
</div>
`;

exports[`MarkdownViewer render when markdown is empty should not render anything 1`] = `""`;

exports[`MarkdownViewer render when not valid html should not fail 1`] = `
<div
  className="markdown-viewer"
>
  <div
    key="0"
  />
</div>
`;

exports[`MarkdownViewer render when passing html true prop should render customComponents nested inside html 1`] = `
<div
  className="markdown-viewer"
>
  <div
    key="0"
  >
    <ItemList
      items={
        Array [
          "items1",
        ]
      }
      key="0"
      link="/library/param1"
      options={
        Object {
          "option1": "optionValue",
        }
      }
    />
    <div
      key="1"
    >
      <MarkdownMedia
        config="(url)"
        key="0"
      />
    </div>
  </div>
</div>
`;

exports[`MarkdownViewer render when passing html true prop should render html 1`] = `
<div
  className="markdown-viewer"
>
  <div
    key="0"
  >
    <h1
      key="0"
    >
      test
    </h1>
  </div>
</div>
`;
